### PR TITLE
feat: update extended client:visible to use an object instead of a string

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

Revert #9363 and redo it using an object instead of a string, that way it can supports options we'll add in the future.

## Testing

Tested manually, like the original PR.

## Docs

Will update docs!

### CODE REVIEW

File: packages/astro/src/runtime/client/visible.ts [hunk 1]
```diff
@@ -11,8 +11,11 @@ const visibleDirective: ClientDirective = (load, options, el) => {
 		await hydrate();
 	};
 
-	const ioOptions = {
-		rootMargin: typeof options.value === 'string' ? options.value : undefined,
+	const rawOptions =
+		typeof options.value === 'object' ? (options.value as ClientVisibleOptions) : undefined;
```
Comment (natemoo-re) [inline packages/astro/src/runtime/client/visible.ts:15]: Just checking, do we already use `JSON.stringify`/`JSON.parse` on `options.value`? If not, we probably get a string here that needs to be deserialized.

**Review by natemoo-re — COMMENTED** [review]
Looks good! Can we add a test that these arguments are properly encoded in the HTML?

File: packages/astro/src/runtime/client/visible.ts (hunk 1 above)
Comment (Princesseuh) [reply packages/astro/src/runtime/client/visible.ts:15]: We do! I tested the change locally and everything works. I'm not sure where we do that, though. I'll add a E2E test.. somewhere

File: packages/astro/e2e/custom-client-directives.test.js [hunk 2]
```diff
@@ -89,4 +89,13 @@ function testClientDirectivesShared() {
 		// Hydrated, this should be 1
 		await expect(counterValue).toHaveText('1');
 	});
+
+	test('Client directives should be passed options correctly', async ({ astro, page }) => {
```
Comment (Princesseuh) [inline packages/astro/e2e/custom-client-directives.test.js:93]: This doesn't test it for `client:visible` exactly... But still, it test that options are passed correctly to client directives (custom ones too, in this case)

File: packages/astro/e2e/fixtures/custom-client-directives/client-options.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// Hydrate on first click on the window
+export default async (load, options) => {
+		const hydrate = await load()
+		document.write(`<div id="options">Passed options are: ${JSON.stringify(options.value)}</div>`)
```
Comment (natemoo-re) [inline packages/astro/e2e/fixtures/custom-client-directives/client-options.js:4]: Lmao I have not seen a `document.write` in quite a while!

**Review by natemoo-re — APPROVED** [review]
Nice test! Thank you!

File: packages/astro/src/runtime/client/visible.ts (hunk 1 above)
Comment (natemoo-re) [inline packages/astro/src/runtime/client/visible.ts:15]: Final nit: this _technically_ adds support for all `IntersectionObserverInit` values, but our types are narrowed to only support `Pick<IntersectionObserverInit, 'rootMargin'>`. Is that desired?

Comment (Princesseuh) [reply packages/astro/src/runtime/client/visible.ts:15]: It doesn't, see the code below

**Pull request merged**